Draft for addition of ELASTIC_STABLESPREAD/USDC as a price identifier (#106)
UMAprotocol/UMIPs: UMIPs/umip-29.md

## Headers

> | UMIP-28     |                                                                                                                                          |

Reviewer: Can you change the title to UMIP-30 and the filename to umip-30? #102 will be UMIP-29 due to the timing of the PR submission

|------------|------------------------------------------------------------------------------------------------------------------------------------------|

> | UMIP Title | Add ELASTIC_STABLESPREAD-USDC as a price identifier              |

Reviewer 2: Hey @chicfilabae we discussed naming conventions as a team, and first of all we apologize that we don't have any documentation about that yet (soon to come!). The problem with this name is that the "-" between the base (STABLESPREAD) and quote (USDC) is inconsistent with other identifier names that have no chars between the base and quote (e.g. USDETH, BCHNBTC, etc.). 

However, "ELASTIC_STABLESPREADUSDC" is practically unreadable, so we've come up with the following rule:

- X = 5 (TBD what this value should be)
- If either the base or the quote currency name is > X chars, then use a "/" to separate the base and quote currency names, otherwise have no separator

Would you mind making this change throughout the UMIP please?

Reviewer: And within the other similar UMIPs!

> | UMIP Title | Add ELASTIC_STABLESPREAD-USDC as a price identifier              |

Reviewer 2: ```suggestion
| UMIP Title | Add ELASTIC_STABLESPREAD/USDC as a price identifier              |
```

Author: Resolved.

| Authors    | Bae (bae@youmychicfila.com), K (k@youmychicfila.com) |
| Status     | Draft                                                                                                                                    |
| Created    | December 23, 2020                                                                                                                           |
 
## Summary (2-5 sentences)

> The DVM should support price requests for the ELASTIC_STABLESPREAD price index. ELASTIC_STABLESPREAD is defined as: `min(max(A - B + 1, 0), 2)`, where `A` refers to an equally weighted basket of {ESD, FRAX, BAC}. B is `MUSD`. ESD is Empty Set Dollar, an algorithmic stablecoin on Ethereum with a dynamic supply model. FRAX is a fractional-algorithmic stablecoin on Ethereum that operates on a creative approach to maintaining a peg via a partially collateralized approach. BAC is Basis Cash, a dynamic supply stablecoin on Ethereum, essentially a version of what the original Basis project meant to be a few years ago. `MUSD` is an autonomous and non-custodial stablecoin on mStable, representing a basket of stablecoins on Ethereum, namely USDT, USDC, TUSD, and DAI. 

Author: FRAX is much newer than DSD but has done a remarkably better job of maintaining its peg than DSD, which largely is susceptible to whale game-theoretic PnDs (a la Cool Rick) to capitalize on cooperating/defecting given the low liquidity. 

W.r.t. BAC vs ESD- BAC has shares and bonds to regulate dollar target, while ESD just has ESD. Different mechanics slightly

> The DVM should support price requests for the ELASTIC_STABLESPREAD-USDC price index, denominated in USDC. ELASTIC_STABLESPREAD-USDC is defined as: `min(max(A - B + 1, 0), 2)`, where `A` refers to an equally weighted basket of {ESD, FRAX, BAC}. B is `MUSD`. ESD is Empty Set Dollar, an algorithmic stablecoin on Ethereum with a dynamic supply model. FRAX is a fractional-algorithmic stablecoin on Ethereum that operates on a creative approach to maintaining a peg via a partially collateralized approach. BAC is Basis Cash, a dynamic supply stablecoin on Ethereum, essentially a version of what the original Basis project meant to be a few years ago. `MUSD` is an autonomous and non-custodial stablecoin on mStable, representing a basket of stablecoins on Ethereum, namely USDT, USDC, TUSD, and DAI. 

Reviewer 2: ```suggestion
The DVM should support price requests for the ELASTIC_STABLESPREAD-USDC price index, which denominates the ELASTIC_STABLESPREAD value in USDC. ELASTIC_STABLESPREAD-USDC is defined as: `min(max(A - B + 1, 0), 2)`, where `A` refers to an equally weighted basket of {ESD, FRAX, BAC}. B is `MUSD`. ESD is Empty Set Dollar, an algorithmic stablecoin on Ethereum with a dynamic supply model. FRAX is a fractional-algorithmic stablecoin on Ethereum that operates on a creative approach to maintaining a peg via a partially collateralized approach. BAC is Basis Cash, a dynamic supply stablecoin on Ethereum, essentially a version of what the original Basis project meant to be a few years ago. `MUSD` is an autonomous and non-custodial stablecoin on mStable, representing a basket of stablecoins on Ethereum, namely USDT, USDC, TUSD, and DAI. 
```

> The DVM should support price requests for the ELASTIC_STABLESPREAD-USDC price index, denominated in USDC. ELASTIC_STABLESPREAD-USDC is defined as: `min(max(A - B + 1, 0), 2)`, where `A` refers to an equally weighted basket of {ESD, FRAX, BAC}. B is `MUSD`. ESD is Empty Set Dollar, an algorithmic stablecoin on Ethereum with a dynamic supply model. FRAX is a fractional-algorithmic stablecoin on Ethereum that operates on a creative approach to maintaining a peg via a partially collateralized approach. BAC is Basis Cash, a dynamic supply stablecoin on Ethereum, essentially a version of what the original Basis project meant to be a few years ago. `MUSD` is an autonomous and non-custodial stablecoin on mStable, representing a basket of stablecoins on Ethereum, namely USDT, USDC, TUSD, and DAI. 

Reviewer 2: ```suggestion
The DVM should support price requests for the ELASTIC_STABLESPREAD/USDC price index, denominated in USDC. ELASTIC_STABLESPREAD/USDC is defined as: `min(max(A - B + 1, 0), 2)`, where `A` refers to an equally weighted basket of {ESD, FRAX, BAC}. B is `MUSD`. ESD is Empty Set Dollar, an algorithmic stablecoin on Ethereum with a dynamic supply model. FRAX is a fractional-algorithmic stablecoin on Ethereum that operates on a creative approach to maintaining a peg via a partially collateralized approach. BAC is Basis Cash, a dynamic supply stablecoin on Ethereum, essentially a version of what the original Basis project meant to be a few years ago. `MUSD` is an autonomous and non-custodial stablecoin on mStable, representing a basket of stablecoins on Ethereum, namely USDT, USDC, TUSD, and DAI. 
```

Let's walk through some cases to better understand what happens in various cases for ELASTIC_STABLESPREAD. If both of these baskets were to trade at their theoretical value, then the resulting value of ELASTIC_STABLESPREAD will be 1. If the Ethereum based stablecoin basket loses its peg and appreciates in value relative to the non-Ethereum stablecoin basket, then the value of ELASTIC_STABLESPREAD will be less than 1. If the opposite happens, then it will be more than 1. It is important that this only measures "relative ability to maintain peg". That is, if both of them lose their peg in the same form, the resulting value still may be 1. 
 

> There is little cost associated with adding this price identifier, as there are multiple free and easily accessible data sources available.

Reviewer: This is a good point. Unnecessary language, but also don't think it's bad to reiterate either.

 
## Technical Specification
The definition of this identifier should be:
 
- Identifier name: ELASTIC_STABLESPREAD-USDC

> - Base Currency: ELASTIC_STABLESPREAD-USDC

Reviewer: Quote currency should be USDC and Base should be ELASTIC_STABLESPREAD

> - Base Currency: ELASTIC_STABLESPREAD/USDC

Reviewer: Base Currency is ELASTIC_STABLESPREAD
Quote Currency would be USDC

- Data Sources: {Uniswap V2: FRAX/ETH} for FRAX, {Uniswap V2: ESD/ETH} for ESD, {Uniswap v2: BAC/ETH} for BAC, {Balancer: MUSD/USDC, Uniswap V2: MUSD/USDC} for MUSD, {Uniswap v2: USDC/ETH} for ETH
- Result Processing: For each constituent asset of the basket, the average of exchanges. Normalizing the ETH result to USDC for FRAX, ESD, and BAC.
- Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.

> - Decimals: 8

Reviewer: The decimal amount will need to be less than the collateral currency that this will be used with. Since USDC has 6, this should be specified to <6. @nicholaspai @chrismaree can you confirm this?

Author: Okay, modifying. I think Matt said something similar to me on the call as well.

Reviewer 2: I don't think this is the case anymore since we now assume that all resolved prices can be specified up to 18 decimals of precision, but still good practice.

- Rounding: Closest, 0.5 up

> - Pricing Interval: 60 seconds

Reviewer: Pricing interval should change to 1 second similar to the other umips

## Implementation
 
The value of ELASTIC_STABLESPREAD-USDC for a given timestamp can be determined with the following process.
 
1. ESD, FRAX, BAC, MUSD should be queried for from the exchanges listed in the "Technical Specification" section for the given timestamp rounded to the nearest second. The results of these queries should be kept at the level of precision they are returned at.
2. For each one, the average of the prices should be calculated.
3. Then, calculate 1/3 * ESD + 1/3 * FRAX + 1/3 * BAC, denote this as `A`, which is in ETH terms

> 4. Perform A * ETH/USDC - MUSD + 1

Reviewer: Shouldn't the ETH/USDC conversion be done after step 6, similar to in https://github.com/UMAprotocol/UMIPs/pull/109/files. Otherwise this would always return 2.

Reviewer 2: Yeah let's keep them standard

Author: I flipped it, it should be USDC/ETH and not ETH/USDC. The intent of the PR you linked to is different- this one is just normalizing since the most liquid pairs are ETH based.

Reviewer 2: Ah good point

5. Perform the maximum of the result of step 4 and 0. This is to ensure non-negativity. 
6. Perform the minimum of the result of step 5 and 2. This is to ensure symmetry.

> 7. This result should be rounded to eight decimal places.

Reviewer: Should be 6 to match the technical implementation section

 
Additionally, [CryptoWatch API](https://docs.cryptowat.ch/rest-api/) is a useful reference. This feed should be used as a convenient way to query the price in realtime, but should not be used as a canonical source of truth for voters. Users are encouraged to build their own offchain price feeds that depend on other sources.
 
Voters are responsible for determining if the result of this process differs from broad market consensus. This is meant to be vague as $UMA tokenholders are responsible for defining broad market consensus.
 
This is only a reference implementation, how one queries the exchanges should be varied and determined by the voter to ensure that there is no central point of failure.

> As an example of how one may go about constructing such a feed programmatically, here is one that was put together as a standalone script with minimal dependencies: https://gist.github.com/chicfilabae/88b5805164d9d5347b9f72b371c237cb. As always, double-check the implementation as well as include additional measures for reliability to ensure reasonable assumptions around when the graph or cryptowatch isn't available. 

Reviewer 3: ```suggestion
As an example of how one may go about constructing such a feed programmatically, here is one that was put together as a standalone script with minimal dependencies which can be found [here](https://gist.github.com/chicfilabae/88b5805164d9d5347b9f72b371c237cb). As always, double-check the implementation as well as include additional measures for reliability to ensure reasonable assumptions around when the graph or cryptowatch isn't available. 
```

Author: Updated
